File: MediatorPattern/MediatorDemo.cs

class User
{
    private int id;
    private string name;
    private Dictionary<int, User> friends;

    public User(int id, string name)
    {
        this.id = id;
        this.name = name;
        friends = new Dictionary<int, User>();
    }

    private void SendMessage(string message)
    {
        Console.WriteLine($"{name} got a message: {message}");
    }

    public void AddFriend(User friend)
    {
        friends.Add(friend.id, friend);
    }

    public void MessageAll(string message)
    {
        foreach (var entry in friends)
            entry.Value.SendMessage($"[ From: {name} ] {message}");
    }

    public void MessageOne(int id, string message)
    {
        if (friends.TryGetValue(id, out User friend))
            friend.SendMessage($"[ From: {name} ] {message}");
    }
}

class MediatorDemo
{
    static void Main(string[] args)
    {
        var eric = new User(1, "Eric Gamma");
        var john = new User(2, "John Vlissides");
        var ralph = new User(3, "Ralph Johnson");
        var richard = new User(4, "Richard Helm");

        eric.AddFriend(john);
        eric.AddFriend(ralph);
        eric.AddFriend(richard);

        john.AddFriend(eric);
        john.AddFriend(ralph);
        john.AddFriend(richard);

        ralph.AddFriend(eric);
        ralph.AddFriend(john);
        ralph.AddFriend(richard);

        richard.AddFriend(eric);
        richard.AddFriend(john);
        richard.AddFriend(ralph);

        eric.MessageAll("Hey everybody!");
        ralph.MessageOne(1, "Hi Eric!");
    }
}

File: MediatorPattern/MediatorDemo.ts
class User {
    private userList = new Map<number, User>();

    public constructor(
        private id: number,
        private name: string
    ) { }

    addFriend(friend: User) {
        this.userList.set(friend.id, friend);
    }

    messageAll(message: string) {
        for (let entry of this.userList.entries()) {
            entry[1].sendMessage(`From [${this.name}]: ${message}`);
        }
    }

    messageOne(id: number, message: string) {
        let user = this.userList.get(id);

        if (null != user) {
            user.sendMessage(`From [${this.name}] ${message}`);
        }
    }

    sendMessage(msg: string) {
        console.log(`[${this.name}] got a message ${msg}`);
    }
}

let eric = new User(1, "Eric Gamma");
let john = new User(2, "John Vlissides");
let ralph = new User(3, "Ralph Johnson");
let richard = new User(4, "Richard Helm");

eric.addFriend(john);
eric.addFriend(ralph);
eric.addFriend(richard);

john.addFriend(eric);
john.addFriend(ralph);
john.addFriend(richard);

ralph.addFriend(eric);
ralph.addFriend(john);
ralph.addFriend(richard);

richard.addFriend(eric);
richard.addFriend(john);
richard.addFriend(ralph);

eric.messageAll("Hey everybody");
ralph.messageOne(1, "Hi Eric");
